python/lsst/meas/astrom/pessimistic_pattern_matcher_b_3D.py

```diff
@@ -0,0 +1,722 @@
+from __future__ import division, print_function, absolute_import
+
+import numpy as np
+from scipy.spatial import cKDTree
+
+
+class PessimisticPatternMatcherB(object):
+    """Class implimenting a pessimistic version of Optimsitic Pattern Matcher B
+    from Tabur 2007. The class loads and stores the reference object in a
+    convienent data structure for matching any set of source obejcts that are
+    assumed to contain each other. The pessimistic nature of the algorithm
+    comes from requiring that it discovers at least two patterns that agree on
+    the correct shift and rotation for matching before exiting. The original
+    behavior of OPMb can be recovered simply.
+    ----------------------------------------------------------------------------
+    Attributes:
+        reference_catalog: Input array of spherical points x, y, z of to use as
+        reference objects for pattern matching.
+    """
+
+    def __init__(self, reference_catalog, log):
+
+        self._reference_catalog = reference_catalog
+        self._n_reference = len(self._reference_catalog)
+        self.log = log
+
+        self._build_distances_and_angles()
+
+    def _build_distances_and_angles(self):
+        """ Function to create the data structures we will use to search
+        for our match in.
+        """
+
+        self._id_array = np.empty(
```

Comment:
Please document each data structure somewhere -- perhaps using inline comments where the structure is created. Since they are private variables, that probably makes more sense that adding them to this doc string.